core/drivers/bnxt/bnxt_fw.c

```diff
@@ -0,0 +1,684 @@
+				   uintptr_t dst,
+				   uint32_t entries,
+				   int inc_src_addr)
+{
+	uint32_t copied_entries;
+
+	while (entries) {
+		copied_entries = bnxt_write32_multiple(dst, src, entries,
+						       inc_src_addr);
+
+		if (copied_entries < entries) {
+			dst += copied_entries * sizeof(uint32_t);
+			src += (inc_src_addr) ?
+				(copied_entries * sizeof(uint32_t)) : 0;
+			entries -= copied_entries;
+		} else {
+			entries = 0;
+		}
+	}
+}
+
+static uint32_t ape_host_view_addr_get(uint32_t bnxt_view_addr, uint32_t size)
+{
+	struct ape_mem_region_s *region = ape_mem_regions;
+	uint32_t addr = 0;
+
+	for ( ; region->size != 0; region++) {
+		if (bnxt_view_addr < region->c_base)
+			continue;
+
+		if (bnxt_view_addr >= (region->c_base + region->size))
+			continue;
+
+		if (size > (region->c_base + region->size - bnxt_view_addr)) {
+			EMSG("ERROR: 0x%x + 0x%x spans memory boundary\n",
+			     bnxt_view_addr, size);
+			break;
+		}
+
+		addr = bnxt_view_addr - region->c_base;
+		addr += region->h_base;
+		break;
+	}
+
+	return addr;
+}
+
+static uint32_t ape_hdr_crc_calc(const struct ape_bin_hdr_s *hdr)
+{
+	uint32_t crc;
+	uint32_t dummy = 0;
+
+	/* Compute the CRC up to, but not including, the checksum field */
+	crc = CRC32(CRC32_INIT_VAL,
+		    (const char *)hdr,
+		    (uintptr_t)(&hdr->checksum) - (uintptr_t)hdr);
+
+	/* Compute the CRC with the checksum field zeroed out */
+	crc = CRC32(~crc, (const char *)&dummy, sizeof(uint32_t));
+
+	/*
+	 * Compute the remainder part of the image header, i.e., the
+	 * section headers
+	 */
+	crc = CRC32(~crc,
+		    (const char *)((uintptr_t)hdr + APE_BIN_HDR_SIZE),
+		    hdr->num_total_sections * APE_SECTION_HDR_SIZE);
+
+	return crc;
+}
+
+static int ape_bin_hdr_valid(const struct ape_bin_hdr_s *hdr)
+{
+	uint32_t checksum;
+
+	if (!hdr) {
+		EMSG("ERROR: no APE image header\n");
+		return BNXT_FAILURE;
+	}
+
+	if (hdr->signature != APE_BIN_HDR_SIGNATURE) {
+		EMSG("ERROR: bad APE image signature\n");
+		return BNXT_FAILURE;
+	}
+
+	if (hdr->num_total_sections > APE_IMG_MAX_SECTIONS) {
+		EMSG("ERROR: too many sections in APE image\n");
+		return BNXT_FAILURE;
+	}
+
+	checksum = ape_hdr_crc_calc(hdr);
+	if (hdr->checksum != checksum) {
+		EMSG("ERROR: bad APE header checksum (exp: %x, act: %x)\n",
+		     hdr->checksum, checksum);
+		return BNXT_FAILURE;
+	}
+
+	return BNXT_SUCCESS;
+}
+
+static int get_char(uint8_t *inbuf, size_t *inbuf_idx, size_t inbuf_size)
+{
+	int c;
+
+	if (*inbuf_idx >= inbuf_size)
+		return EOF;
+
+	c = inbuf[*inbuf_idx];
+	*inbuf_idx += 1;
+
+	return c;
+}
+
+static void put_char(uint8_t *outbuf,
+		     size_t *outbuf_idx,
+		     size_t outbuf_size,
+		     uint8_t ch)
+{
+	if (*outbuf_idx >= outbuf_size)
+		return;
+
+	outbuf[*outbuf_idx] = ch;
+	*outbuf_idx += 1;
+}
+
+static size_t ape_section_uncompress(uint8_t *inbuf,
+				     size_t inbuf_size,
+				     uint8_t *outbuf,
+				     size_t outbuf_size)
+{
+	int i, j, k, r, c;
```

Comment:
All local variables should normally be initialized with a value. The rationale is that in case some variable isn't assigned another value later we still have a rather defined behaviour and we're not giving away junk from the stack.

I'm only commenting in this function, but the comment applies to the entire pull request.

Reply:
Will have initialized local variables here and in other functions